### PR TITLE
+per #17837 Detect (and repair) corrupt event streams

### PR DESCRIPTION
**Pull request by patriknw**

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #17837

**Comment by patriknw** [comment]
@ktoso I have done a refactoring of how plugin config is handled, since we want most things to be configured per plugin, also replay-filter, see ce9edf4

File: akka-persistence/src/main/resources/reference.conf [hunk 1]
```diff
@@ -82,11 +78,64 @@ akka.persistence {
...
+
+      # Dispatcher for the plugin actor.
+      plugin-dispatcher = "akka.persistence.dispatchers.default-plugin-dispatcher"
+      
+      # Dispatcher for message replay.
+      replay-dispatcher = "akka.persistence.dispatchers.default-replay-dispatcher"
+      
+      # Set this to true if plugin actor has a constructor which expects plugin 
+      # configuration entry.
+      inject-config = false
```
Comment (ktoso) [inline akka-persistence/src/main/resources/reference.conf]: why as an option? We usually "try to find the right constructor"

File: akka-persistence/src/main/resources/reference.conf [hunk 2]
```diff
@@ -82,11 +78,64 @@ akka.persistence {
...
+        call-timeout = 10s
+        reset-timeout = 30s
+      }
+      
+      # The replay filter can detect a corrupt event stream by inspecting 
+      # sequence numbers and writerUuid when replaying events. 
+      replay-filter {
+        # What the filter should do when detecting invalid events.
+        # Supported values:
+        # `repair` : discard events from old writers, warning is logged
```
Comment (ktoso) [inline akka-persistence/src/main/resources/reference.conf]: maybe `discard-old`?

File: akka-persistence/src/main/resources/reference.conf (hunk 1 above)
Comment (patriknw) [inline akka-persistence/src/main/resources/reference.conf]: good point, let's create ticket for that

File: akka-persistence/src/main/resources/reference.conf (hunk 2 above)
Comment (patriknw) [inline akka-persistence/src/main/resources/reference.conf]: `repair-by-discard-old` ?

Comment (ktoso) [inline akka-persistence/src/main/resources/reference.conf]: Think we could have different repair strategies? How about `repair-*`, so `repair-discard-old`

File: akka-persistence/src/main/scala/akka/persistence/journal/ReplayFilter.scala [hunk 3]
```diff
@@ -0,0 +1,155 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.persistence.journal
+
+import akka.actor.ActorRef
+import akka.actor.Actor
+import akka.persistence.JournalProtocol
+import java.util.LinkedList
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/journal/ReplayFilter.scala]: <3 :)

File: akka-persistence/src/main/resources/reference.conf (hunk 2 above)
Comment (patriknw) [inline akka-persistence/src/main/resources/reference.conf]: yes, good point

File: akka-persistence/src/test/scala/akka/persistence/PersistentActorFailureSpec.scala [hunk 4]
```diff
@@ -288,6 +288,34 @@ class PersistentActorFailureSpec extends PersistenceSpec(PersistenceSpec.config(
...
+      p2 ! Cmd("c")
+      p2 ! GetState
+      expectMsg(List("a-1", "a-2", "c-1", "c-2"))
+
+      // Create yet another one with same persistenceId, b-1 and b-2 discarded during replay
+      EventFilter.warning(start = "Invalid replayed event", occurrences = 2) intercept {
+        val p3 = namedPersistentActor[Behavior1PersistentActor]
+        p3 ! GetState
+        expectMsg(List("a-1", "a-2", "c-1", "c-2"))
+      }
```
Comment (ktoso) [inline akka-persistence/src/test/scala/akka/persistence/PersistentActorFailureSpec.scala]: great :)

File: akka-persistence/src/main/scala/akka/persistence/journal/ReplayFilter.scala [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+  def sendBuffered(): Unit = {
+    val iter = buffer.iterator()
+    while (iter.hasNext())
+      persistentActor.tell(iter.next(), Actor.noSender)
+    buffer.clear()
+  }
+
+  def logIssue(errMsg: String): Unit = mode match {
+    case Warn | Repair ⇒ log.warning(errMsg)
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/journal/ReplayFilter.scala]: Good, at least people will know something fishy happened

**Comment by ktoso** [comment]
LGTM after mode rename; Please create a ticket for https://github.com/akka/akka/pull/18195#discussion_r37066295

File: akka-persistence/src/main/resources/reference.conf (hunk 1 above)
Comment (patriknw) [inline akka-persistence/src/main/resources/reference.conf]: #18219

**Pull request merged**